hushline/premium.py

```diff
@@ -0,0 +1,445 @@
+        stripe_invoice.amount_remaining = invoice.amount_remaining
+        stripe_invoice.status = StripeInvoiceStatusEnum(invoice.status)
+        db.session.commit()
+
+        # Invoice has been paid, so make sure the user is upgraded
+        user = db.session.query(User).filter_by(stripe_subscription_id=invoice.subscription).first()
+        if user:
+            user.tier_id = BUSINESS_TIER
+            db.session.commit()
+    else:
+        raise ValueError(f"Could not find invoice with ID {invoice.id}")
+
+
+async def worker(app: Flask) -> None:
+    current_app.logger.error("Starting worker")
+    with app.app_context():
+        while True:
+            stripe_event = (
+                db.session.query(StripeEvent)
+                .filter_by(status="pending")
+                .order_by(StripeEvent.created_at)
+                .first()
+            )
+            if not stripe_event:
+                await asyncio.sleep(10)
+                continue
+
+            stripe_event.status = "in_progress"
+            db.session.add(stripe_event)
+            db.session.commit()
+
+            event_json = json.loads(stripe_event.event_data)
+            event = stripe.Event.construct_from(event_json, current_app.config["STRIPE_SECRET_KEY"])
+
+            current_app.logger.info(
+                f"Processing event {stripe_event.event_type} ({stripe_event.event_id})"
+            )
+            try:
+                # subscription events
+                if event.type.startswith("customer.subscription."):
+                    subscription: stripe.Subscription = stripe.Subscription.construct_from(
+                        event.data.object, current_app.config["STRIPE_SECRET_KEY"]
+                    )
+                    if event.type == "customer.subscription.created":
+                        handle_subscription_created(subscription)
+                    elif event.type == "customer.subscription.updated":
+                        handle_subscription_updated(subscription)
+                    elif event.type == "customer.subscription.deleted":
+                        handle_subscription_deleted(subscription)
+                # invoice events
+                elif event.type.startswith("invoice."):
+                    invoice: stripe.Invoice = stripe.Invoice.construct_from(
+                        event.data.object, current_app.config["STRIPE_SECRET_KEY"]
+                    )
+                    if event.type == "invoice.created":
+                        handle_invoice_created(invoice)
+                    elif event.type == "invoice.payment_succeeded":
+                        handle_invoice_payment_succeeded(invoice)
+
+            except Exception as e:
+                current_app.logger.error(
+                    f"Error processing event {stripe_event.event_type} ({stripe_event.event_id}): {e}\n{stripe_event.event_data}"  # noqa: E501
+                )
+                stripe_event.status = "error"
```

Comment:
should we save more error detail with the event?

Reply:
It now saved error messages too, in https://github.com/scidsg/hushline/pull/595/commits/217c78adaa08b5863f9cf699e2af415904f8c8b5